agent/resource.go

```diff
@@ -187,15 +204,65 @@ func (a *Agent) processIncomingDeleteResourceRequest(ctx context.Context, req *e
 
 // rreqEventToGvk returns a GroupVersionResource object, name and namespace for
 // the resource requested in rreq.
-func rreqEventToGvk(rreq *event.ResourceRequest) (gvr schema.GroupVersionResource, name string, namespace string) {
+func rreqEventToGvk(rreq *event.ResourceRequest) (gvr schema.GroupVersionResource, name string, namespace string, subresource string) {
 	gvr.Group = rreq.Group
 	gvr.Resource = rreq.Resource
 	gvr.Version = rreq.Version
 	name = rreq.Name
 	namespace = rreq.Namespace
+	subresource = rreq.Subresource
 	return
 }
 
+// getManagedResourceSubresource retrieves a subresource of a single resource from the cluster
+// and returns it as unstructured data. The resource to be returned must be managed by Argo CD.
+func (a *Agent) getManagedResourceSubresource(ctx context.Context, gvr schema.GroupVersionResource, name, namespace, subresource string) (*unstructured.Unstructured, error) {
+	// First check if the parent resource is managed
+	_, err := a.getManagedResource(ctx, gvr, name, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	// If parent is managed, fetch the subresource using a direct REST call
+	// The dynamic client doesn't directly support subresources, so we build the path manually
+	var path string
+	if namespace != "" {
+		if gvr.Group == "" {
+			// Core API: /api/v1/namespaces/{namespace}/{resource}/{name}/{subresource}
+			path = fmt.Sprintf("/api/%s/namespaces/%s/%s/%s/%s", gvr.Version, namespace, gvr.Resource, name, subresource)
+		} else {
+			// Named group API: /apis/{group}/{version}/namespaces/{namespace}/{resource}/{name}/{subresource}
+			path = fmt.Sprintf("/apis/%s/%s/namespaces/%s/%s/%s/%s", gvr.Group, gvr.Version, namespace, gvr.Resource, name, subresource)
+		}
+	} else {
+		if gvr.Group == "" {
+			// Core API cluster-scoped: /api/v1/{resource}/{name}/{subresource}
+			path = fmt.Sprintf("/api/%s/%s/%s/%s", gvr.Version, gvr.Resource, name, subresource)
+		} else {
+			// Named group API cluster-scoped: /apis/{group}/{version}/{resource}/{name}/{subresource}
+			path = fmt.Sprintf("/apis/%s/%s/%s/%s/%s", gvr.Group, gvr.Version, gvr.Resource, name, subresource)
+		}
+	}
+
+	// Use the discovery client's RESTClient to make the request
+	restClient := a.kubeClient.Clientset.Discovery().RESTClient()
+	req := restClient.Get().AbsPath(path)
+
+	// Execute the request
+	bytes, err := req.DoRaw(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get subresource %s: %w", subresource, err)
+	}
+
+	// Convert the raw response to unstructured
+	result := &unstructured.Unstructured{}
+	if err := json.Unmarshal(bytes, &result.Object); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal subresource response: %w", err)
+	}
```

Comment:
Doing the request and unmarshaling the results are being repeated throughout all new functions. It would be nice to have less repetition in the code.